src/cargo/util/dependency_queue.rs

```diff
@@ -82,36 +84,39 @@ impl<N: Hash + Eq + Clone, E: Eq + Hash + Clone, V> DependencyQueue<N, E, V> {
     /// All nodes have been added, calculate some internal metadata and prepare
     /// for `dequeue`.
     pub fn queue_finished(&mut self) {
+        let mut out = HashMap::new();
         for key in self.dep_map.keys() {
-            depth(key, &self.reverse_dep_map, &mut self.depth);
+            depth(key, &self.reverse_dep_map, &mut out);
         }
+        self.priority = out.into_iter().map(|(n, set)| (n, set.len())).collect();
 
-        fn depth<N: Hash + Eq + Clone, E: Hash + Eq + Clone>(
+        fn depth<N: Hash + Eq + Clone + std::fmt::Debug + PartialEq, E: Hash + Eq + Clone>(
             key: &N,
             map: &HashMap<N, HashMap<E, HashSet<N>>>,
-            results: &mut HashMap<N, usize>,
-        ) -> usize {
-            const IN_PROGRESS: usize = !0;
-
-            if let Some(&depth) = results.get(key) {
-                assert_ne!(depth, IN_PROGRESS, "cycle in DependencyQueue");
-                return depth;
+            results: &mut HashMap<N, HashSet<N>>,
+        ) -> HashSet<N> {
+            let in_progress: HashSet<N> = HashSet::new();
+            if let Some(depth) = results.get(key) {
+                assert_ne!(depth, &in_progress, "cycle in DependencyQueue");
+                return depth.clone();
             }
+            results.insert(key.clone(), in_progress);
 
-            results.insert(key.clone(), IN_PROGRESS);
+            let mut set = HashSet::new();
+            set.insert(key.clone());
 
-            let depth = 1 + map
+            for dep in map
                 .get(key)
                 .into_iter()
                 .flat_map(|it| it.values())
                 .flat_map(|set| set)
-                .map(|dep| depth(dep, map, results))
-                .max()
-                .unwrap_or(0);
+            {
+                set.extend(depth(dep, map, results).into_iter())
+            }
 
-            *results.get_mut(key).unwrap() = depth;
+            *results.get_mut(key).unwrap() = set.clone();
```

Comment:
Could this perhaps avoid `.clone()` by returning `&HashSet<N>` inside the `HashMap` itself?

Reply:
as in, `depth` returns a `&HashSet<N>`

Reply:
WIth witch lifetime? I tried `'results`, but it did not go well. Is there something else worth trying? Is it worth `Rc<HashSet<N>>`?

Reply:
I was thinking of a diff like:

```diff
diff --git a/src/cargo/util/dependency_queue.rs b/src/cargo/util/dependency_queue.rs
index 21886c0ee..946549507 100644
--- a/src/cargo/util/dependency_queue.rs
+++ b/src/cargo/util/dependency_queue.rs
@@ -88,14 +88,15 @@ impl<N: Hash + Eq + Clone, E: Eq + Hash + Clone, V> DependencyQueue<N, E, V> {
         }
         self.priority = out.into_iter().map(|(n, set)| (n, set.len())).collect();
 
-        fn depth<N: Hash + Eq + Clone, E: Hash + Eq + Clone>(
+        fn depth<'a, N: Hash + Eq + Clone, E: Hash + Eq + Clone>(
             key: &N,
             map: &HashMap<N, HashMap<E, HashSet<N>>>,
-            results: &mut HashMap<N, HashSet<N>>,
-        ) -> HashSet<N> {
-            if let Some(depth) = results.get(key) {
+            results: &'a mut HashMap<N, HashSet<N>>,
+        ) -> &'a HashSet<N> {
+            if results.contains_key(key) {
+                let depth = &results[key];
                 assert!(!depth.is_empty(), "cycle in DependencyQueue");
-                return depth.clone();
+                return depth;
             }
             results.insert(key.clone(), HashSet::new());
 
@@ -108,12 +109,12 @@ impl<N: Hash + Eq + Clone, E: Eq + Hash + Clone, V> DependencyQueue<N, E, V> {
                 .flat_map(|it| it.values())
                 .flat_map(|set| set)
             {
-                set.extend(depth(dep, map, results))
+                set.extend(depth(dep, map, results).iter().cloned())
             }
 
-            *results.get_mut(key).unwrap() = set.clone();
-
-            set
+            let slot = results.get_mut(key).unwrap();
+            *slot = set;
+            return &*slot;
         }
     }
 ```

if that feels to complicated though it's definitely too minor to use `Rc`, so we can just go w/ what's here as-is.

Reply:
You are a wizard! Hat's off!